x/crosschain/types/cctx.go

```diff
@@ -0,0 +1,189 @@
+package types
+
+import (
+	"fmt"
+
+	cosmoserrors "cosmossdk.io/errors"
+	sdkmath "cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	observertypes "github.com/zeta-chain/zetacore/x/observer/types"
+)
+
+// GetCurrentOutTxParam returns the current outbound tx params.
+// There can only be one active outtx.
+// OutboundTxParams[0] is the original outtx, if it reverts, then
+// OutboundTxParams[1] is the new outtx.
+func (m CrossChainTx) GetCurrentOutTxParam() *OutboundTxParams {
+	if len(m.OutboundTxParams) == 0 {
+		return &OutboundTxParams{}
+	}
+	return m.OutboundTxParams[len(m.OutboundTxParams)-1]
+}
+
+// IsCurrentOutTxRevert returns true if the current outbound tx is the revert tx.
+func (m CrossChainTx) IsCurrentOutTxRevert() bool {
+	return len(m.OutboundTxParams) == 2
+}
+
+// OriginalDestinationChainID returns the original destination of the outbound tx, reverted or not
+// If there is no outbound tx, return -1
+func (m CrossChainTx) OriginalDestinationChainID() int64 {
+	if len(m.OutboundTxParams) == 0 {
+		return -1
+	}
+	return m.OutboundTxParams[0].ReceiverChainId
+}
+
+// GetAllAuthzZetaclientTxTypes returns all the authz types for zetaclient
+
+func (m CrossChainTx) Validate() error {
+	if m.InboundTxParams == nil {
+		return fmt.Errorf("inbound tx params cannot be nil")
+	}
+	if m.OutboundTxParams == nil {
+		return fmt.Errorf("outbound tx params cannot be nil")
+	}
+	if m.CctxStatus == nil {
+		return fmt.Errorf("cctx status cannot be nil")
+	}
+	if len(m.OutboundTxParams) > 2 {
+		return fmt.Errorf("outbound tx params cannot be more than 2")
+	}
+	if m.Index != "" {
+		err := ValidateZetaIndex(m.Index)
+		if err != nil {
+			return err
+		}
+	}
+	err := m.InboundTxParams.Validate()
+	if err != nil {
+		return err
+	}
+	for _, outboundTxParam := range m.OutboundTxParams {
+		err = outboundTxParam.Validate()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+/*
+AddRevertOutbound does the following things in one function:
+
+	1. create a new OutboundTxParams for the revert
+
+	2. append the new OutboundTxParams to the current OutboundTxParams
+
+	3. update the TxFinalizationStatus of the current OutboundTxParams to Executed.
+*/
+
+func (m *CrossChainTx) AddRevertOutbound(gasLimit uint64) {
```

Comment:
Hmm, it would be caught in the cctx.Validate  (moved it here so that we run validate irrespective of success or fail )
https://github.com/zeta-chain/zeta-node/blob/b79405d7ce1c87622797c5c69143bc8872271099/x/crosschain/keeper/msg_server_vote_outbound_tx.go#L259-L262 

But I see no hard in adding a check , 
b67126a276c3a88f779c2d77460743b01a2fd260